File: JavaScript/Data Structures/LinkedList/LinkedList.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
    }

    insert(value) {
        if (this.head === null) {
            this.head = new Node(value);
        } else {
            let newNode = new Node(value);
            newNode.next = this.head;
            this.head = newNode;
        }
    }

    removeFromFront() {
        let removedNode = this.head;
        this.head = this.head.next;
        removedNode.next = null;
        return removedNode;
    }

    empty() {
        return !this.head
    }

    print() {
        let curNode = this.head;
        while (curNode) {
            console.log(curNode.value);
            curNode = curNode.next;
        }
    }

    getNode(n) {
        let i = 0;
        let curNode = this.head;
        while(curNode && i < n) {
            i += 1;
            curNode = curNode.next;
        }

        return curNode
    }
}

export { Node, LinkedList };

File: JavaScript/PracticeProblems/LinkedList/LinkedList.go
package linkedlist

import "fmt"

// Node in LinkedList.
type Node struct {
	value	int
	next	*Node
}

// LinkedList struct that we will use to create methods.
type LinkedList struct {
	head	*Node
}

func (l *LinkedList) insert(v int) {
	if l.head == nil {
		l.head = &Node{value: v, next: nil}
	} else {
		n := Node{value: v, next: l.head}
		l.head = &n
	}
}

// remove
func (l *LinkedList) remove(v int) *Node {
	var c = l.head
	var r *Node

	if c.value == v {
		r = l.head
		l.head = l.head.next
		return r
	}
	
	for c != nil {
		if c.next.value == v {
			r = c.next
			c.next = c.next.next
			return r
		}
		c = c.next
	}

	return nil
}

// print
func (l *LinkedList) print() {
	var c = l.head

	for c != nil {
		fmt.Println(c.value)
		c = c.next
	}
}

// find
func (l *LinkedList) find(v int) *Node {
	var c = l.head

	for c != nil {
		if c.value == v {
			return c
		}
		c = c.next
	}

	return nil
}

// removeFront
func (l *LinkedList) removeFront() *Node {
	f := l.head
	l.head = l.head.next
	return f
}

// func main() {
// 	linkedList := LinkedList{head: nil}
// 	linkedList.insert(1)
// 	linkedList.insert(2)
// 	linkedList.insert(3)
// 	f := linkedList.find(2)
// 	fmt.Println(*f)
// 	fmt.Println("--")
// 	n := linkedList.remove(2)
// 	fmt.Println(*n)
// 	fmt.Println("--")
// 	linkedList.print()
// 	fmt.Println("--")
// 	f2 := linkedList.removeFront()
// 	fmt.Println(*f2)
// 	//fmt.Println("hello world")
// }